pinot-core/src/main/java/org/apache/pinot/core/util/trace/TraceContext.java

```diff
@@ -45,9 +49,28 @@
  * request from tracing to prevent resource leak.
  */
 public final class TraceContext {
+
+  static Map<Long, Pair<QueryContext, Future[]>> requestId2FuturesMap = new ConcurrentHashMap<>();
+
   private TraceContext() {
   }
 
+  public static void register(long requestId, QueryContext queryContext, Future[] futures) {
+    requestId2FuturesMap.put(requestId, Pair.of(queryContext, futures));
+  }
+
+  public static void cancel(long requestId) {
+    Pair<QueryContext, Future[]> queryContextFuturesPair = requestId2FuturesMap.get(requestId);
```

Comment:
use remove() instead? to avoid keeping futures around. 
alternatively, can also clean things up in unregister() method.